File: desktop-client/src/app/account/services/notification/notification.service.ts
import { Injectable, OnDestroy } from '@angular/core';
import { FriendRequestResponse } from '@app/account/interfaces/friend-request-response';
import { SoundService } from '@app/shared/services/sound/sound.service';
import { WebsocketService } from '@app/shared/services/websocket/websocket.service';
import { FRIEND_NOTIFICATION_SOUND } from '@common/constants/account-sound';
import { NOT_FOUND_INDEX } from '@common/constants/verif-constants';
import { Notification } from '@common/interfaces/notification';
import { FriendEvent } from '@common/socket-events/friend-event';
import { NotificationEvent } from '@common/socket-events/notification-event';

@Injectable({
    providedIn: 'root',
})
export class NotificationService implements OnDestroy {
    notificationList: Notification[] = [];
    show: boolean = false;

    constructor(
        private readonly websocketService: WebsocketService,
        private readonly soundService: SoundService,
    ) {
        this.websocketService.on(NotificationEvent.ADD_NOTIFICATION, (notification?: Notification) => {
            if (notification) {
                this.addNotification(notification);
                this.soundService.playSound(FRIEND_NOTIFICATION_SOUND);
            }
        });
        this.getNotificationsEvent();
    }

    toggleShow(isOff: boolean = false) {
        if (isOff) this.show = false;
        else this.show = !this.show;
    }

    ngOnDestroy() {
        this.websocketService.removeAllListeners(NotificationEvent.ADD_NOTIFICATION);
    }

    async getNotificationsEvent() {
        const notifications = await this.websocketService.sendWithAck<undefined, Notification[]>(NotificationEvent.GET_NOTIFICATIONS);
        if (notifications) this.notificationList = notifications;
    }

    sendFriendRequest(username: string) {
        this.websocketService.send<string>(NotificationEvent.SEND_FRIEND_REQUEST, username);
    }

    respondToFriendRequest(username: string, isAccepted: boolean, notificationId: string) {
        const payload: FriendRequestResponse = { senderName: username, isAccepted, notificationId };
        this.websocketService.send<FriendRequestResponse>(FriendEvent.FRIEND_REQUEST_RESPONSE, payload);
        this.removeNotification(notificationId);
    }

    private removeNotification(notificationId: string) {
        const index = this.notificationList.findIndex((notification) => notification.id === notificationId);
        if (index > NOT_FOUND_INDEX) {
            this.notificationList.splice(index, 1);
        }
    }

    private addNotification(notification: Notification) {
        this.notificationList.push(notification);
    }
}

File: mobile_client/lib/services/notification_service.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:mobile_client/constants/account_notification.dart';
import 'package:mobile_client/constants/socket_events/friend_event.dart';
import 'package:mobile_client/constants/socket_events/notification_event.dart';
import 'package:mobile_client/models/friend_request.dart';
import 'package:mobile_client/models/notification.dart';
import 'package:mobile_client/services/sound_service.dart';
import 'package:mobile_client/services/websocket_service.dart';

class NotificationService extends ChangeNotifier {
  final WebSocketService _webSocketService;
  final SoundService _soundService;

  List<Notification> notificationList = [];

  NotificationService(this._webSocketService, this._soundService);

  void initialize() {
    _webSocketService.on<dynamic>(NotificationEvent.addNotification.name,
        (data) {
      final notification = Notification.fromJson(data);
      addNotification(notification);
    });
    getNotificationsEvent();
  }

  Future<void> getNotificationsEvent() async {
    final data = await _webSocketService.sendWithAck<void, List<dynamic>>(
        NotificationEvent.getNotification.name);
    notificationList = data
        .map((notification) => Notification.fromJson(notification))
        .toList();
    notifyListeners();
  }

  void respondToFriendRequest(
      Notification notification, bool isAccepted) async {
    final answer = FriendRequestResponse(
      senderName: notification.senderUsername,
      isAccepted: isAccepted,
      notificationId: notification.id,
    );
    _webSocketService.send<FriendRequestResponse>(
        FriendEvent.friendRequestResponse.name, answer);
    removeNotification(notification.id);
  }

  void removeNotification(String notificationId) {
    notificationList
        .removeWhere((notification) => notification.id == notificationId);
    notifyListeners();
  }

  void addNotification(Notification notification) {
    notificationList.add(notification);
    _soundService.playSound(friendNotificationSound);
    notifyListeners();
  }
}
